### PR TITLE
Test harness for Move transactional tests

### PR DESCRIPTION
**Pull request by jjleng**
### Description
Enables Move developers to write end-to-end tests in the Aptos repo. Compared to smoke tests, transactional tests can be validated on developers' workstations. They are much faster to run :)

### Test Plan
cargo test -p aptos-transactional-test-harness

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2022)

### CODE REVIEW

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (areshand) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:744]: I previously had this issue https://github.com/aptos-labs/aptos-core/issues/535. 
It would be ideal if AptosTestAdapter is just a wrapper around Aptos VM adapter without depending on the  MoveTestAdapter.  It would be easier when the move upgrade or our Aptos VM adpater upgrades
Not sure how complex this will be.

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (areshand) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:47]: @vgao1996 should we directly fork move_transactional_test_runner to Aptos Repo. 
For example. Our test should rely on Aptos Adapter instead of movetest adapter which is not aware aptos specific changes.

Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:47]: Good point. I also want to customize InitCommand and move quickly.

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:744]: I will look into the issue you reported. Getting rid of MoveTestAdapter should not be very hard, but we might need to maintain the copied/cloned code separately. Don't know if we still want to keep track of the upstream changes in case upstream has some useful updates. Sui is also using MoveTestAdapter https://github.com/MystenLabs/sui/blob/8628837f3a3dd4fc6cb5ca5dea4a90b13e926e74/crates/sui-transactional-test-runner/src/test_adapter.rs#L25. Maybe we can live with this for now. I need to spend time on improving the ergonomics of this test harness first.

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move
Comment (areshand) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move:35]: this is very cool. 
just want to know your thought on supporting writing assert directly in the move file with ==, >=, <=, > and < operators. 
For example, 
//# view --address Alice --resource Alice::HelloWorld::ModuleData  --field_name="global_counter" --comparator=larger_than --target_value=0

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:180]: maybe just
```
/**
 * Parsing
 */
```

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/publish_module.move
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/publish_module.move:6]: update to public entry

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move:11]: update casing

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:879]: maybe make more effective in whitespace

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:560]: what happens if you use the compiled version?

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move:5]: humor me, does this framework support struct types?

**Review by davidiw — COMMENTED** [review]
Overall looks pretty solid. A couple of questions:
* What are the types supported by the test syntax -- can it take structs? Or will this be a labor of love from us?
* We'll need to build the right testing mentality across (unit tests, transactional tests, e2e tests, and smoke tests) -- would love to get your thoughts there about what each one uniquely solves.
* My gut feeling is that we'll need to figure out
  * How to integrate this into the CLI
  * How to make the e2e-testsuite a library so that others can write their own tests

FWIW, I may have combined the concept of transactional tests and e2e-testsuite in my mind. It might be good to just brainstorm for a little bit of time on testing strategies and requirements -- to answer the question where tests belong and making sure that those can run as fast as possible.

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move:35]: Can you elaborate more on how to do it through the Move script? But we can get that working in a separate PR.

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move:11]: Coin and TestCoin cases stayed the same after https://github.com/aptos-labs/aptos-core/pull/2018

Comment (wrwg) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/execute_script.move:11]: #2032 changes this and lands soon.

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move
Comment (areshand) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/call_function.move]: I am also curious how to do this assertion in move. If I have a move script after `view --address Alice --resource Alice::HelloWorld::ModuleData`. This move script won't be able to access the struct fields defined other module. We probably can only parse the resource

**Review by davidiw — COMMENTED** [review]
Overall happy to move this along... one favor to ask... can we add more tests here? Especially around tables and more complex structs / resources? Maybe even move some of the smoke tests for tokens, multi-agent, and strings to here? If those work here, then we're golden to move forward!

**Review by areshand — APPROVED** [review]

**Comment by davidiw** [comment]
Reviewing now...

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/table.move
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/table.move:21]: not bad, not bad at all

how do we learn the handle though?

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (davidiw) [inline aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:376]: We should definitely be using the types instead of this: https://github.com/aptos-labs/aptos-core/tree/main/types/src/account_config/resources

same for the resource above

**Review by davidiw — COMMENTED** [review]
also before we land, can we also have some sort of readme to go along with this?

**Review by wrwg — APPROVED** [review]
I think we should land this and then refactor or remove it. 

- The `e2etestsuite` is not well working either because it is based on the language once upon a time called Move -- the MoveIR.
- The transactional tests are missing programmatic assertions and a pure Rust interface.

We need some merge of the ideas of both, but perhaps this is done best when both approaches are in use.

**Comment by wrwg** [comment]
> I think we should land this and then refactor or remove it.
> 
> * The `e2etestsuite` is not well working either because it is based on the language once upon a time called Move -- the MoveIR.
> * The transactional tests are missing programmatic assertions and a pure Rust interface.
> 
> We need some merge of the ideas of both, but perhaps this is done best when both approaches are in use.

... forget to mention, approved indeed only after other reviewers requests are addressed.

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:376]: updated

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:420]: This function is only used to fetch the native coin, which is the gas coin. It won't be exposed to get any coin balances

File: aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/table.move
Comment (jjleng) [reply aptos-move/aptos-transactional-test-harness/tests/aptos_test_harness/table.move:21]: The able handles can be found in the output of previous command `//# view --address Alice  --resource 0x1::token::Collections`

**Pull request merged**
